pkg/typesystem/typesystem.go

```diff
@@ -764,32 +764,30 @@ func (t *TypeSystem) RecursiveTTUCanFastPathV2(objectType, relation, userType st
 				// if the edge does not have a weight for the terminal type, we can skip it
 				continue
 			}
-			// edge is a set operator thus we have to inspect each node of the operator
-			if edge.GetEdgeType() == graph.RewriteEdge {
-				// if the operator node has weight infinite we need to get all the edges to evaluate the preconditions
-				if w == graph.Infinite {
+			if w == graph.Infinite {
+				// edge is a set operator thus we have to inspect each node of the operator
+				if edge.GetEdgeType() == graph.RewriteEdge {
+					// if the operator node has weight infinite we need to get all the edges to evaluate the preconditions
 					operationalEdges, okOpEdge := t.authzWeightedGraph.GetEdgesFromNode(edge.GetTo())
 					if !okOpEdge {
 						return false
 					}
 					// these edges will need to be evaluated in subsequent iterations
 					innerEdges = append(innerEdges, operationalEdges...)
 					continue
-				}
-			}
-			// find and validate the TTUEdge which is infinite (the one being processed at the current time)
-			if edge.GetEdgeType() == graph.TTUEdge &&
-				edge.GetTuplesetRelation() == tuplesetRelation && strings.HasSuffix(edge.GetTo().GetUniqueLabel(), computedRelationSuffix) {
-				// The recursive TTU edge needs to have:
-				// 1. weight infinite for the terminal type
-				// 2. the computed relation for the TTU is the same relation
-				// 3. there can only be one recursive TTU edge
-				if w == graph.Infinite && edge.GetTo() == objRelNode && !recursiveTTUFound {
+				} else if edge.GetEdgeType() == graph.TTUEdge && !recursiveTTUFound && edge.GetTuplesetRelation() == tuplesetRelation && strings.HasSuffix(edge.GetTo().GetUniqueLabel(), computedRelationSuffix) && edge.GetTo() == objRelNode {
+					// find and validate the TTUEdge which is infinite (the one being processed at the current time)
+					// The recursive TTU edge needs to have:
+					// 1. weight infinite for the terminal type
+					// 2. the computed relation for the TTU is the same relation
+					// 3. there can only be one recursive TTU edge
 					recursiveTTUFound = true
 					continue
 				}
 			}
+
 			// catch all, everything has to be weight 1 regardless of direct, computed, rewrite.
+			// thus if an infinite didn't get handled it will exit through here
 			if w > 1 {
```

Comment:
```suggestion
			else if w > 1 {
```
nitpik: if w is infinite, we don't need to evaluate for w again.

Reply:
Left comment explicitly for this. 

> // thus if an infinite didn't get handled it will exit through here

If it doesn't get handled by the cases of infinity above, it needs to exit. For example, a TTU would be handled there. Or when `recursiveTTUFound` is already `true` (aka you have the same TTU in 2 different relations).

Reply:
Ah...thanks!